internal/testrunner/runners/system/servicedeployer/custom_agent.go

```diff
@@ -0,0 +1,198 @@
+func NewCustomAgentDeployer(cfgPath string) (*CustomAgentDeployer, error) {
+	path, err := createCustomAgentYaml(cfgPath)
+	if err != nil {
+		return nil, err
+	}
+
+	return &CustomAgentDeployer{
+		cfg: path,
+	}, nil
+}
+
+// SetUp sets up the service and returns any relevant information.
+func (d *CustomAgentDeployer) SetUp(inCtxt ServiceContext) (DeployedService, error) {
+	logger.Debug("setting up service using Docker Compose service deployer")
+
+	appConfig, err := install.Configuration()
+	if err != nil {
+		return nil, errors.Wrap(err, "can't read application configuration")
+	}
+
+	env := append(
+		appConfig.StackImageRefs(install.DefaultStackVersion).AsEnv(),
+		fmt.Sprintf("%s=%s", serviceLogsDirEnv, inCtxt.Logs.Folder.Local),
+	)
+
+	service := dockerComposeDeployedService{
+		ymlPaths: []string{d.cfg},
+		project:  "elastic-package-service",
+		sv: ServiceVariant{
+			Name: dockerCustomAgentName,
+			Env:  env,
+		},
+	}
+
+	outCtxt := inCtxt
+
+	p, err := compose.NewProject(service.project, service.ymlPaths...)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not create Docker Compose project for service")
+	}
+
+	// Verify the Elastic stack network
+	err = stack.EnsureStackNetworkUp()
+	if err != nil {
+		return nil, errors.Wrap(err, "Elastic stack network is not ready")
+	}
+
+	// Clean service logs
+	err = files.RemoveContent(outCtxt.Logs.Folder.Local)
+	if err != nil {
+		return nil, errors.Wrap(err, "removing service logs failed")
+	}
+
+	inCtxt.Name = dockerCustomAgentName
+	serviceName := inCtxt.Name
+	opts := compose.CommandOptions{
+		Env:       env,
+		ExtraArgs: []string{"--build", "-d"},
+	}
+	err = p.Up(opts)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not boot up service using Docker Compose")
+	}
+
+	// Connect service network with stack network (for the purpose of metrics collection)
+	err = docker.ConnectToNetwork(p.ContainerName(serviceName), stack.Network())
+	if err != nil {
+		return nil, errors.Wrapf(err, "can't attach service container to the stack network")
+	}
+
+	err = p.WaitForHealthy(opts)
+	if err != nil {
+		processServiceContainerLogs(p, compose.CommandOptions{
+			Env: opts.Env,
+		}, outCtxt.Name)
+		return nil, errors.Wrap(err, "service is unhealthy")
+	}
+
+	// Build service container name
+	outCtxt.Hostname = p.ContainerName(serviceName)
+
+	logger.Debugf("adding service container %s internal ports to context", p.ContainerName(serviceName))
+	serviceComposeConfig, err := p.Config(compose.CommandOptions{Env: env})
+	if err != nil {
+		return nil, errors.Wrap(err, "could not get Docker Compose configuration for service")
+	}
+
+	s := serviceComposeConfig.Services[serviceName]
+	outCtxt.Ports = make([]int, len(s.Ports))
+	for idx, port := range s.Ports {
+		outCtxt.Ports[idx] = port.InternalPort
+	}
+
+	// Shortcut to first port for convenience
+	if len(outCtxt.Ports) > 0 {
+		outCtxt.Port = outCtxt.Ports[0]
+	}
```

Comment:
nit: I don't remember if this condition is still required.

Reply:
if we keep it then the port can be referenced from the test config and that can be useful in some cases, not a hard requirement though so we can remove it if there is any concern